File: guests/haskell/split.hs
{-# LANGUAGE OverloadedStrings #-}

-- The split scene, Haskell port — guests/rust/split.rs,
-- tools/scenes/split.steps.

import Data.Word (Word64)
import Data.Text (Text)
import qualified Data.Text as T
import KayaApp

detailId :: Word64
detailId = 7

main :: IO ()
main = kayaMain $ \app -> do
  _ <- buildTx app $ do
    window primary [WTitle "split", WPanes 2]
    s <- signal (T.pack "list pane")
    root <-
      column
        []
        [ -- Authored ids so the REAL-TREE read can address these: an
          -- index read passes whether or not anything reached the
          -- screen.
          labelBound s [A11yId ("list" :: Text)], -- label#0
          buttonOn "open detail"
            ( buildTx app $ do
                pushEntry
                  detailId
                  [ ETitle "detail",
                    EOnPopped (buildTx app (writeSignal s (T.pack "popped detail")))
                  ]
                caption <- signal (T.pack "detail pane")
                pane <- column [] [labelBound caption [A11yId ("detail" :: Text)]]
                mountIn detailId pane
            )
            []
        ]
    mount root
    return s
  return ()

File: guests/rust/split.rs
//! The split scene, driven by TWO scripts (tools/scenes/split.steps and
//! listdetail.steps): the presentation is asked for ONCE and never again.

/// `split` drives the size class with real resizes; `listdetail` asserts
/// the bare invariant at whatever width its host gives.
pub(crate) fn app(ctx: kaya::AppCtx) {
    use kaya::WindowId;

    #[derive(Clone, Copy)]
    enum Msg {
        OpenDetail,
        PoppedDetail,
    }

    const DETAIL: WindowId = WindowId(7);

    let msgs = kaya::Messages::<Msg>::new();
    let status = ctx.apply(|tx| {
        tx.window(kaya::DEFAULT_WINDOW)
            .title("split")
            .panes(2);
        let status = tx.signal("list pane");
        let root = tx
            .column(|tx| {
                // Authored ids: an index read passes for an empty arm.
                tx.label(status).a11y_id("list"); // label#0
                let open = tx.button("open detail").id(); // button#0
                msgs.on_click(open, Msg::OpenDetail);
            })
            .id();
        tx.mount(root);
        status
    });

    while let Some(msg) = msgs.next(&ctx) {
        match msg {
            Msg::OpenDetail => {
                let entry = ctx.apply(|tx| {
                    let entry = tx.push_entry(DETAIL).title("detail").id();
                    let pane = tx
                        .column(|tx| {
                            let caption = tx.signal("detail pane");
                            tx.label(caption).a11y_id("detail");
                        })
                        .id();
                    tx.mount_in(entry, pane);
                    entry
                });
                msgs.on_entry_popped(entry, Msg::PoppedDetail);
            }
            // Retention: the base root took this write while covered.
            Msg::PoppedDetail => ctx.apply(|tx| {
                tx.write(status, "popped detail");
            }),
        }
    }
}

fn main() {
    kaya::run(app)
}
